File: Collections/OneToOneOrderedFreezableDictionary.vb
Public Class OneToOneOrderedFreezableDictionary(Of TKey, TValue)
    Inherits OneToOneOrderedDictionary(Of TKey, TValue)

    Public Sub New(ByVal KeySelector As Func(Of TValue, TKey))
        MyBase.New(KeySelector)
    End Sub

#Region "Freezable"
    Public Sub Freeze()
        If Me.FreezeCalled Then
            Exit Sub
        End If
        Me.FreezeCalled = True
        Me.OnFreezing()
        Me._IsFreezed = True
    End Sub

    Protected Overridable Sub OnFreezing()

    End Sub

    Protected Sub VerifyWrite()
        Verify.False(Me._IsFreezed, "Cannot change a freezed object.")
    End Sub

#Region "IsFreezed Property"
    Private _IsFreezed As Boolean

    Public ReadOnly Property IsFreezed As Boolean
        Get
            Return Me._IsFreezed
        End Get
    End Property
#End Region

    Private FreezeCalled As Boolean
#End Region

    Public Overrides Sub Add(Value As TValue)
        Me.VerifyWrite()
        MyBase.Add(Value)
    End Sub

    Public Overrides Function [Set](Value As TValue) As Boolean
        Me.VerifyWrite()
        Return MyBase.Set(Value)
    End Function

    Public Overrides Sub Clear()
        Me.VerifyWrite()
        MyBase.Clear()
    End Sub

    Public Overrides Sub Insert(Index As Integer, Value As TValue)
        Me.VerifyWrite()
        MyBase.Insert(Index, Value)
    End Sub

    Public Overrides Sub RemoveAt(index As Integer)
        Me.VerifyWrite()
        MyBase.RemoveAt(index)
    End Sub

    Public Overrides Function RemoveKey(key As TKey) As Boolean
        Me.VerifyWrite()
        Return MyBase.RemoveKey(key)
    End Function

End Class

File: Collections/OneToOneOrderedFreezableDictionary.cs
using System;

namespace Ks
{
    namespace Common
    {
        public class OneToOneOrderedFreezableDictionary<TKey, TValue> : OneToOneOrderedDictionary<TKey, TValue>
        {
            public OneToOneOrderedFreezableDictionary(Func<TValue, TKey> KeySelector) : base(KeySelector)
            {
            }


            public void Freeze()
            {
                if (this.FreezeCalled)
                    return;
                this.FreezeCalled = true;
                this.OnFreezing();
                this._IsFreezed = true;
                this.OnFroze();
            }

            protected virtual void OnFreezing()
            {
            }

            protected virtual void OnFroze()
            {
            }

            protected void VerifyWrite()
            {
                Verify.False(this._IsFreezed, "Cannot change a freezed object.");
            }

            private bool _IsFreezed;

            public bool IsFrozen
            {
                get
                {
                    return this._IsFreezed;
                }
            }

            private bool FreezeCalled;

            public override TValue this[int index]
            {
                get
                {
                    return base[index];
                }
                set
                {
                    this.VerifyWrite();
                    base[index] = value;
                }
            }

            public override bool Set(TValue Value)
            {
                this.VerifyWrite();
                return base.Set(Value);
            }

            public override void Clear()
            {
                this.VerifyWrite();
                base.Clear();
            }

            public override void Insert(int Index, TValue Value)
            {
                this.VerifyWrite();
                base.Insert(Index, Value);
            }

            public override void RemoveAt(int index)
            {
                this.VerifyWrite();
                base.RemoveAt(index);
            }

            public override bool RemoveKey(TKey key)
            {
                this.VerifyWrite();
                return base.RemoveKey(key);
            }
        }
    }
}
